Add LedSC integration

PatrikKr010 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Added new requirement: [websc-client==1.0.1](https://pypi.org/project/websc-client/).

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

PatrikKr010 committed:
Add LedSC integration

PatrikKr010 committed:
The code formatted using Ruff

PatrikKr010 committed:
Add docstrings and  clean code.

PatrikKr010 committed:
update manifest and call script.gen_requirements_all

PatrikKr010 committed:
Merge branch 'dev' into dev

PatrikKr010 committed:
LedSC: Update documentation link

PatrikKr010 committed:
Merge remote-tracking branch 'origin/dev' into dev

PatrikKr010 committed:
Merge branch 'dev' into dev

PatrikKr010 commented:
[Documentation PR](https://github.com/home-assistant/home-assistant.io/pull/32648)

PatrikKr010 committed:
Merge branch 'dev' into dev

PatrikKr010 committed:
Merge branch 'dev' into dev

PatrikKr010 committed:
Merge branch 'dev' into dev

tr4nt0r reviewed (changes requested):
An Integration may not communicate directly with devices or services as stated in the development checklist:
https://developers.home-assistant.io/docs/development_checklist/

PatrikKr010 commented:
Hi @tr4nt0r,
Communication is done using websocket. The pythoni library `requests`, which is available in PyPi, is used for this. This library has been added to requirements.

I wrote the WebSC application with which the communication takes place in the interaction. This application is used to remove the hardware part and transfer it to the websocket API.

I can separate the integration part into a separate library that will take care of the communication with WebSC, but I find it redundant. 

Is this separation necessary?

tr4nt0r commented:
> Hi @tr4nt0r, Communication is done using websocket. The pythoni library `requests`, which is available in PyPi, is used for this. This library has been added to requirements.
> 
> I wrote the WebSC application with which the communication takes place in the interaction. This application is used to remove the hardware part and transfer it to the websocket API.
> 
> I can separate the integration part into a separate library that will take care of the communication with WebSC, but I find it redundant.
> 
> Is this separation necessary?

requests is already part of Home Assistant, so it's quite redundant to add it to the manifest but anyway, the maintainers are pretty strict about this and requests is simply not considered a device specific abstraction layer, it is just an http client

PatrikKr010 committed:
Merge branch 'dev' into dev

PatrikKr010 committed:
Added abstraction layer for communication with WebSC

PatrikKr010 commented:
Hi @tr4nt0r 
I added an abstract layer that handles communication with WebSC Websocket API. This layer is the [websc-client](https://pypi.org/project/websc-client/) library. I hope this abstraction is OK.

tr4nt0r commented on homeassistant/components/ledsc/light.py:
I suggest you add an EntityDescription to your light entities, (import it from homeassistant.components.light,) then you can make use translation strings and icons for your entities. Also you should add DeviceInfo attribute to your entities.

tr4nt0r commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+            try:
+                info = await validate_input(self.hass, user_input)
+            except ConnectionError:
+                errors["base"] = "Cannot connect"
```
This is not the error message that is displayed, it is a tranlation key and the corresponding message is looked up in the `strings.json` file. Please add a `strings.json` file to your integration.
```suggestion
                errors["base"] = "cannot_connect
```

PatrikKr010 replied:
Done

tr4nt0r commented on homeassistant/components/ledsc/light.py:
```diff
@@ -0,0 +1,175 @@
+        return not self._websc.is_lost
+
+    @property
+    def name(self) -> str:
```
Use the entity_description attribute instead and use strings.json for device names instead

PatrikKr010 replied:
I can use `entity_description` for more detailed device information, but the entity name is dynamic based on the WebSC configuration.

tr4nt0r replied:
I think you don't even need a name for the entity, if you leave it empty, and set the name of the device and the manufacturer in the DeviceInfo it will use the generic name "Light", the entity will then be named light.ledsc_devicename_light

PatrikKr010 replied:
![image](https://github.com/home-assistant/core/assets/57815256/44950f2d-3ff5-4f70-abcc-3884c9a1db5b)

Do is not a very nice default name.
If you don't mind, I would like to leave this property. This name is set in WebSC.
![Snímek obrazovky z 2024-05-30 18-33-19](https://github.com/home-assistant/core/assets/57815256/4358baf6-59bb-4832-82c6-b9f0731cd2bc)

tr4nt0r replied:
It's tricky to get this right, haven't done this yet with the light platform. I would like to test it, is there some way to do this? The host demo.ledsc.eu did not work.

PatrikKr010 replied:
The demo.ledsc.eu domain is still under development. I am currently creating a way to create a virtual instance of WebSC so that anyone who wants to try the demo can run their own instance of WebSC for a limited time. However, for testing purposes I can run one static instance on this domain. Otherwise you can run your own instances of course.

PatrikKr010 replied:
You can try: http://demo.ledsc.eu:8080/

tr4nt0r replied:
nope, server rejects the connection

tr4nt0r commented on homeassistant/components/ledsc/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "codeowners": ["@PatrikKr010"],
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/ledsc/",
+  "integration_type": "hub",
```
```suggestion
  "integration_type": "device",
```

PatrikKr010 replied:
But device is not the right category. It is really a hub. The integration is connected to a single device using Websocket, but this device controls any number of lights.

The individual devices are configured in the WebSC application. Advanced LedSC properties can be configured in this application. The integration connects to WebSC and loads the configured devices. So it is not necessary to configure the individual devices separately in the integration configuration.

tr4nt0r replied:
you are right

tr4nt0r reviewed (changes requested):
It is required to have tests at least for the config flow. Files that are not covered by tests must be added to `.coveragerc`

PatrikKr010 commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict) -> dict[str, Any]:
+    """
+    Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
```
Personally, I wouldn't add a comment here, but according to the checklist, the code must pass the check using the `python3 -m script.hassfest` command, which requires docstring on methods.
In this state the test passes without any errors.

PatrikKr010 replied:
Done

PatrikKr010 committed:
Use cv.port to validate port number

Co-authored-by: Mr. Bubbles <manni@zapto.de>

PatrikKr010 commented on homeassistant/components/ledsc/light.py:
```diff
@@ -0,0 +1,175 @@
+        """Initialize an AwesomeLight."""
+        self._hass: HomeAssistant = hass
+        self._websc: WebSCAsync = websc
+        self.__id = f"{client_id}-{websc.name}"
+        _LOGGER.info(f"LedSC '%s' initialized!", self.name)
+
+    @property
+    def unique_id(self) -> str | None:
+        """Return id unique for client and entity name combination."""
+        return self.__id
+
```
OK, this log informs about loaded devices that could be useful for debugging. I will change the level of this log to `DEBUG`.

PatrikKr010 replied:
Done

PatrikKr010 committed:
Fix docstrings & code formatting

PatrikKr010 commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+    VERSION = 1
+
+    async def async_step_user(self, user_input: dict[str, Any] | None = None):
+        """Handle the initial step."""
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+
+from .consts import DEFAULT_HOST, DEFAULT_PORT, DOMAIN
+from websc_client import WebSClientAsync as WebSClient
+
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+"""Config flow for Hello World integration."""
```
Done

tr4nt0r reviewed (commented):
> Personally, I wouldn't add a comment here, but according to the checklist, the code must pass the check using the `python3 -m script.hassfest` command, which requires docstring on methods.
In this state the test passes without any errors. 

The first line describing the function is ok, what i meant was the second line.

PatrikKr010 commented on homeassistant/components/ledsc/__init__.py:
```diff
@@ -0,0 +1,43 @@
+"""Platform for light integration."""
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/light.py:
```diff
@@ -0,0 +1,175 @@
+    devices: list[LedSC] = list()
+    for websc in client.devices.values():
+        ledsc = LedSC(
+            client_id=f"{host}:{port}",
```
I used this way of identification to make it unique for the same connection and it was not possible to connect more integrations to the same WebSC server.

Is this ID global or must it be unique only in my integrations? If is global, I suggest f"ledsc:{host}:{port}".

PatrikKr010 replied:
Done

PatrikKr010 committed:
Fix docstrings & code formatting

PatrikKr010 commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Hello World."""
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+
+import voluptuous as vol
+
+from homeassistant import config_entries
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,68 @@
+                    CONF_HOST: user_input[CONF_HOST],
+                    CONF_PORT: user_input[CONF_PORT],
+                }
+            )
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/__init__.py:
```diff
@@ -0,0 +1,43 @@
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PORT): cv.string,
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/__init__.py:
```diff
@@ -0,0 +1,43 @@
+async def options_update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Handle options update."""
+    _LOGGER.debug("Configuration options updated, reloading OneWire integration")
+    await hass.config_entries.async_reload(entry.entry_id)
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/consts.py:
```diff
@@ -0,0 +1,8 @@
+from homeassistant.const import Platform
+
+DOMAIN = "ledsc"
+PLATFORMS: list[str] = [Platform.LIGHT]
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/consts.py:
```diff
@@ -0,0 +1,8 @@
+"""Constant variables."""
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/light.py:
```diff
@@ -0,0 +1,175 @@
+    hass: HomeAssistant, config, add_entities: AddEntitiesCallback, discovery_info=None
+):
+    """Redirects to '__setup'."""
+    hass.async_create_task(__setup(hass, dict(config), add_entities))
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/light.py:
```diff
@@ -0,0 +1,175 @@
+    Connect to WebSC.
+
+    load the configured devices and add them to hass.
+    """
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/light.py:
```diff
@@ -0,0 +1,175 @@
+    await client.connect()
+    hass.async_create_background_task(client.observer(), name="ledsc-observer")
+
+    devices: list[LedSC] = list()
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/light.py:
```diff
@@ -0,0 +1,175 @@
+    host = config["host"]
+    port = config["port"]
+
+    client = WebSClient(host=host, port=port)
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/light.py:
```diff
@@ -0,0 +1,175 @@
+
+
+class LedSC(LightEntity):
+    """Representation of an Awesome Light."""
```
Done

PatrikKr010 committed:
Add untested files to .coveragerc

tr4nt0r reviewed (changes requested):
> > It is required to have tests at least for the config flow. Files that are not covered by tests must be added to `.coveragerc`

Done 

You NEED to add testing for the config flow, it is mandatory for new integrations. If you run scaffold in VS Code and select that you want to add config flow to your integration, it will create some example test files, that should be a good starting point

tr4nt0r commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,60 @@
+        if user_input:
+            self._async_abort_entries_match(user_input)
+            try:
+                info = await validate_input(self.hass, user_input)
+            except ConnectionError:
+                errors["base"] = "cannot connect"
+            else:
```
```suggestion
                info = await validate_input(self.hass, user_input)
            except (ConnectionError, WebSocketException):
                errors["base"] = "cannot connect"
            else:
```
You should also catch `WebSocketException` here

PatrikKr010 replied:
I added this exception class  in [WebSClient](https://gitlab.com/ledsc/websclient/-/commit/c503503884781952d82fe27cd435ccc0491c8224) and create WebSClient custom exception class and except them.

tr4nt0r commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,60 @@
+                errors["base"] = "cannot connect"
+            else:
+                return self.async_create_entry(title=info["title"], data=user_input)
+
+        return self.async_show_form(
+            step_id="user", data_schema=DATA_SCHEMA, errors=errors
+        )
```
```suggestion

        return self.async_show_form(
            step_id="user",
            data_schema=self.add_suggested_values_to_schema(
                data_schema=DATA_SCHEMA,
                suggested_values=self.config_entry.data,
            ),
            errors=errors,
        )
```
The form will otherwise always reset to the default values if an error happens

PatrikKr010 replied:
This object has no attribute 'config_entry'. Tou mean `user_input`?
![Snímek obrazovky z 2024-05-31 09-06-14](https://github.com/home-assistant/core/assets/57815256/dcc70511-d791-4736-9886-5920f5880484)

tr4nt0r replied:
yeah you are right, sorry for the copy/paste error

PatrikKr010 replied:
Done

tr4nt0r commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,60 @@
+            try:
+                info = await validate_input(self.hass, user_input)
+            except ConnectionError:
+                errors["base"] = "cannot connect"
```
```suggestion
                errors["base"] = "cannot_connect"
```
this is a translation key. The corresponding error message should be in strings.json

tr4nt0r commented on homeassistant/components/ledsc/light.py:
```diff
@@ -0,0 +1,153 @@
+
+async def async_setup_entry(
+    hass: HomeAssistant, config: ConfigEntry, add_entities: AddEntitiesCallback
+):
+    """
+    Connect to WebSC.
+
+    load the configured devices from WebSC Server and add them to hass.
+    """
```
```suggestion
):
    """Connect to WebSC.

    load the configured devices from WebSC Server and add them to hass.
    """
```
Ruff complains otherwise

PatrikKr010 replied:
Done

tr4nt0r reviewed (commented):
> I used this way of identification to make it unique for the same connection and it was not possible to connect more integrations to the same WebSC server.

Is this ID global or must it be unique only in my integrations? If is global, I suggest f"ledsc:{host}:{port}". 

Please look at the requirements for unique ids https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements

PatrikKr010 committed:
Set key in error as `cannot_connect`

Co-authored-by: Mr. Bubbles <manni@zapto.de>

PatrikKr010 committed:
Optimize imports & Exceptions & up websc-client to version 1.0.2 & Fix docstrings.

PatrikKr010 commented on homeassistant/components/ledsc/light.py:
```diff
@@ -0,0 +1,153 @@
+from homeassistant.core import HomeAssistant
+
+from websc_client import WebSClientAsync as WebSClient
+from websc_client import WebSCAsync
```
Done

PatrikKr010 committed:
Remove default host in config flow.

PatrikKr010 commented on homeassistant/components/ledsc/consts.py:
```diff
@@ -0,0 +1,5 @@
+"""Constants for the LedSC integration."""
+
+DOMAIN = "ledsc"
+DEFAULT_HOST = "demo.ledsc.eu"
```
Done

tr4nt0r commented on tests/components/ledsc/test_config_flow.py:
```diff
@@ -0,0 +1,76 @@
+            result["flow_id"],
+            USER_INPUT,
+        )
+    assert result2["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
```
```suggestion
    assert result2["type"] is FlowResultType.CREATE_ENTRY
```

PatrikKr010 replied:
Done

tr4nt0r commented on tests/components/ledsc/test_config_flow.py:
```diff
@@ -0,0 +1,76 @@
+    assert result2["data"] == USER_INPUT
+
+
+async def test_form_cannot_connect(hass):
```
```suggestion
async def test_form_cannot_connect(hass) -> None:
```

PatrikKr010 replied:
Done

tr4nt0r commented on tests/components/ledsc/test_config_flow.py:
```diff
@@ -0,0 +1,76 @@
+        context={"source": config_entries.SOURCE_USER},
+    )
+
+    with patch(
+            "homeassistant.components.ledsc.config_flow.validate_input",
+            side_effect=ConnectionError,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
```
```suggestion
    with patch(
            "homeassistant.components.ledsc.config_flow.WebSClient",
            autospe=True,
    ) as mock_websc:
        mock_websc.connect.side_effect = ConnectionError
        result2 = await hass.config_entries.flow.async_configure(
```

Mock out the external library, not code that is part of your integration (please verify first the suggested code works, have not tested it)
Also you should create a fixture out of the library mocking to reduce duplicate code

PatrikKr010 replied:
Done. Test works now

tr4nt0r commented on tests/components/ledsc/test_config_flow.py:
```diff
@@ -0,0 +1,76 @@
+
+async def test_form_unique_id_check(hass):
+    """Test we handle unique id check"""
+    await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": config_entries.SOURCE_USER},
+        data=USER_INPUT,
+    )
```
```suggestion
    MockConfigEntry(domain=DOMAIN, data=USER_INPUT).add_to_hass(hass)
```

You are initializing the config flow 2 times. Just add the config entry before testing the config flow. 
Preferably, create it as fixture in the conftest.py file.

```
@pytest.fixture(name="ledsc_config_entry")
def mock_bring_config_entry() -> MockConfigEntry:
    """Mock ledsc configuration entry."""
    return MockConfigEntry(
        domain=DOMAIN, data= {"host": "127.0.0.1", "port": 8080}
    )
```

PatrikKr010 replied:
But how do I get `flow_id` with this flow initialization structure? Under normal circumstances I get the result of the initialization with all the necessary data.

tr4nt0r commented on tests/components/ledsc/test_config_flow.py:
```diff
@@ -0,0 +1,76 @@
+            "homeassistant.components.ledsc.config_flow.validate_input",
+            side_effect=ConnectionError,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            USER_INPUT,
+        )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "cannot_connect"}
+
```
```suggestion
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            USER_INPUT,
        )

    assert result["type"] == "form"
    assert result["errors"] == {"base": "cannot_connect"}

```

just reuse the result variable

PatrikKr010 replied:
Done

tr4nt0r commented on tests/components/ledsc/test_config_flow.py:
```diff
@@ -0,0 +1,76 @@
+            USER_INPUT,
+        )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "cannot_connect"}
+
+
```
the Config flow should end with FlowResultType.CREATE_ENTRY to show that the config flow is able to recover. So set side_effect back to None, then try again and assert the flow terminated successfully

PatrikKr010 replied:
Done

tr4nt0r commented on tests/components/ledsc/test_config_flow.py:
```diff
@@ -0,0 +1,76 @@
+        context={"source": config_entries.SOURCE_USER},
+        data=USER_INPUT,
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
```
```suggestion
    assert result["type"] is FlowResultType.ABORT
```

PatrikKr010 replied:
Done

tr4nt0r commented on tests/components/ledsc/test_config_flow.py:
```diff
@@ -0,0 +1,76 @@
+from unittest.mock import patch
```
```suggestion
"""Test the LedSc config flow."""

from unittest.mock import patch
```

PatrikKr010 replied:
Done

tr4nt0r commented on .coveragerc:
```diff
@@ -700,6 +700,10 @@ omit =
     homeassistant/components/ld2410_ble/sensor.py
     homeassistant/components/led_ble/__init__.py
     homeassistant/components/led_ble/light.py
+    homeassistant/components/ledsc/__init__.py
+    homeassistant/components/ledsc/light.py
+    homeassistant/components/ledsc/config_flow.py
+    homeassistant/components/ledsc/consts.py
     homeassistant/components/lg_netcast/media_player.py
```
```suggestion
    homeassistant/components/ledsc/__init__.py
    homeassistant/components/ledsc/light.py
    homeassistant/components/lg_netcast/media_player.py
```

PatrikKr010 replied:
Done

tr4nt0r reviewed (approved):
> I used this way of identification to make it unique for the same connection and it was not possible to connect more integrations to the same WebSC server.

Is this ID global or must it be unique only in my integrations? If is global, I suggest f"ledsc:{host}:{port}". 

You have _async_abort_entries_match to make sure, there is only one config entry per ledsc server.  That should do the job.

tr4nt0r reviewed (changes requested):
Your integration is still missing translations. Please add a strings.json file.

PatrikKr010 committed:
Fix tests && Up websc-client to 1.0.3

PatrikKr010 commented on tests/components/ledsc/test_config_flow.py:
```diff
@@ -0,0 +1,76 @@
+ENTRY_CONFIG = {"host": "127.0.0.1", "port": 8080}
+
+
+async def test_form(hass):
```
Done

PatrikKr010 commented on tests/components/ledsc/test_config_flow.py:
```diff
@@ -0,0 +1,76 @@
+from unittest.mock import patch
+
+from homeassistant import config_entries, data_entry_flow
```
Done

PatrikKr010 commented on tests/components/ledsc/test_config_flow.py:
```diff
@@ -0,0 +1,76 @@
+        context={"source": config_entries.SOURCE_USER},
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
```
Done

PatrikKr010 committed:
Update config_flow && light id.

PatrikKr010 committed:
Add strings.json

PatrikKr010 commented:
I fixed the tests. Now all tests are passing. I added a strings.json file, but it doesn't work. But I don't know if I understand its principle correctly.

PatrikKr010 committed:
Merge branch 'dev' into dev

PatrikKr010 committed:
Up websc-client to 1.0.4

PatrikKr010 committed:
Merge branch 'dev' into dev

PatrikKr010 committed:
Merge branch 'dev' into dev

tr4nt0r reviewed (changes requested):
Please address all changes i requested before we can continue with the review process.

tr4nt0r commented:
Also i had the impression you are the manufacturer of these devices or somehow associated with them, or am i wrong? If so, maybe this could be interesting for you:
- https://partner.home-assistant.io/
- https://www.home-assistant.io/blog/2022/07/12/partner-program/

PatrikKr010 committed:
Update homeassistant/components/ledsc/strings.json

Co-authored-by: Mr. Bubbles <manni@zapto.de>

PatrikKr010 committed:
Cleaning & Docs

PatrikKr010 commented on homeassistant/components/ledsc/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+                info = await validate_input(self.hass, user_input)
+            except WebSClientError:
+                errors["base"] = "cannot_connect"
+            else:
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/__init__.py:
```diff
@@ -0,0 +1,35 @@
+import homeassistant.helpers.config_validation as cv
+
+PLATFORMS: list[str] = [Platform.LIGHT]
+_LOGGER = logging.getLogger(__name__)
```
Done

PatrikKr010 commented on homeassistant/components/ledsc/__init__.py:
```diff
@@ -0,0 +1,35 @@
+        vol.Required(CONF_PORT): cv.port,
+    }
+)
+
```
Done

PatrikKr010 commented on tests/components/ledsc/__init__.py:
Done

joostlek commented:
Hey @PatrikKr010,

I just checked your website, ledsc.eu, and apparently I can't order anything. Could you please elaborate on what your plans are and what the product is?

Thanks!

PatrikKr010 commented:
Hello,
Yes, the product cannot be ordered yet. The development and testing is completely finished. We are currently working on the launch of production and arranging the distribution of our product.  
As soon as we have finished this work and the integration into Home assistant is released, we will start selling our product.

joostlek commented:
Thanks for contribution and for the information about the current state of the product. We appreciate the effort you’ve put into developing this integration.

However, we only include integrations for products or services that are currently available for our users. This ensures that our community can immediately benefit from and support the integration in further open source development.

We recommend publishing your integration as a custom component for now. Once your product is available on the market and the custom integration has had time to mature, we would be happy to re-evaluate and review a new pull request for inclusion of your integration in the core repository.

Thank you for understanding, and looking forward to seeing your product’s hitting the market!
